disco_aws_automation/disco_datapipeline.py

```diff
@@ -0,0 +1,255 @@
+        if not pipeline.is_persisted():  # if this is a new pipeline, save the metadata first
+            unique_id = "pipeline_init_%s" % datetime.now().isoformat()  # make the create call idempotent
+            created = throttled_call(self._dp_client.create_pipeline,
+                                     name=pipeline._name, uniqueId=unique_id,
+                                     description=pipeline._description, tags=pipeline._tags or [])
+            pipeline._id = created['pipelineId']
+        # regardless, save the pipeline content:
+        resp = throttled_call(self._dp_client.put_pipeline_definition,
+                              pipelineId=pipeline._id,
+                              pipelineObjects=pipeline._objects or [],
+                              parameterObjects=pipeline._params or [],
+                              parameterValues=pipeline._param_values or [])
+        return resp
+
+    # possibly worth having but not currently needed: set_tags
+    # http://boto3.readthedocs.io/en/latest/reference/services/datapipeline.html#DataPipeline.Client.add_tags
+
+    def fetch(self, pipeline_id):
+        "Fetch a pipeline (metadata and content) from AWS by ID."
+        contents = throttled_call(self._dp_client.get_pipeline_definition, pipelineId=pipeline_id,
+                                  version='latest')
+        meta_resp = throttled_call(self._dp_client.describe_pipelines, pipelineIds=[pipeline_id])
+        meta = meta_resp['pipelineDescriptionList'][0]
+        return AsiaqDataPipeline(
+            pipeline_id=pipeline_id, name=meta['name'], description=meta.get('description'),
+            tags=meta.get('tags'), metadata=meta.get('fields'),
+            contents=contents['pipelineObjects'], parameter_definitions=contents['parameterObjects'],
+            param_values=contents['parameterValues'],
+        )
+
+    def fetch_content(self, pipeline):
+        "Populate the pipeline content fields of this pipeline object with the information fetched from AWS."
+        if pipeline.has_content():
+            raise Exception("Content already fetched or locally generated for this pipeline object")
+        definition = throttled_call(self._dp_client.get_pipeline_definition,
+                                    pipelineId=pipeline._id, version='latest')
+        pipeline.update_content(definition['pipelineObjects'],
+                                definition.get('parameterObjects'),
+                                definition.get('parameterValues'))
+
+    def fetch_all_descriptions(self):
+        """
+        Fetch all pipelines in this account/region, populating only their name/metadata fields
+        (use fetch_content to get the low-level details).
+        """
+        return self.search_descriptions()
+
+    def search_descriptions(self, name=None, tags=None):
+        "Fetch all pipelines in this account/region that have the given tags and/or the given name."
```

Comment:
That's just `fetch`: you can have multiple pipelines with the same name/tags, but the ID is unique.